query-service-impl/src/main/java/org/hypertrace/core/query/service/pinot/PinotBasedRequestHandler.java

```diff
@@ -147,63 +148,110 @@ public QueryCost canHandle(
       }
-    }
 
-    // successfully found a view that can handle the request
-    if (found) {
-      // TODO: Come up with a way to compute the cost based on request and view definition
-      // Higher columns --> Higher cost,
-      // Finer the time granularity --> Higher the cost.
-      cost = 0.5;
+      found = doViewFiltersMatch(analyzer.getOptimizedFilter(), viewFilterMap);
     }
-    QueryCost queryCost = new QueryCost();
-    queryCost.setCost(cost);
-    return queryCost;
+
+    // TODO: Come up with a way to compute the cost based on request and view definition
+    // Higher columns --> Higher cost,
+    // Finer the time granularity --> Higher the cost.
+    return new QueryCost(found ? 0.5 : -1);
   }
 
-  /**
-   * Method to check if the given ViewColumnFilter matches the given query filter. A match here
-   * means the view column is superset of what the query filter is looking for, not an exact match.
-   */
-  private boolean doesFiltersMatch(ViewColumnFilter viewColumnFilter, Filter queryFilter) {
-    if (viewColumnFilter.getOperator() == ViewColumnFilter.Operator.IN) {
-      if (queryFilter.getOperator() != Operator.IN && queryFilter.getOperator() != Operator.EQ) {
+  private boolean doViewFiltersMatch(Filter filter, Map<String, ViewColumnFilter> viewFilterMap) {
+    // Navigate the entire filter tree to check if every query filter which has the
+    // view filter columns has all matching view filters. If not, return false since
+    // this view can't handle such query.
+    if (filter.getChildFilterCount() == 0) {
+      if (viewFilterMap.size() != 1) {
         return false;
       }
 
-      return isSubSet(viewColumnFilter.getValues(), queryFilter.getRhs());
-    } else {
-      // query filter must be an equals filter in this case.
-      if (queryFilter.getOperator() != Operator.EQ) {
+      // If the column names of query filter and view filter match but the filters
+      // don't match, we can exit here.
+      String viewFilterColumnName = viewFilterMap.keySet().stream().findFirst().get();
+      if (filter.getLhs().getValueCase() == ValueCase.COLUMNIDENTIFIER &&
+          StringUtils.equals(viewFilterColumnName,
+              filter.getLhs().getColumnIdentifier().getColumnName()))
+      {
+        return doesSingleViewFilterMatchLeafQueryFilter(viewFilterMap.get(viewFilterColumnName),
+            filter);
+      }
+
+      return false;
+    }
+
+    Map<String, Filter> childLeafFilterMap = filter.getChildFilterList().stream()
+        .filter(f -> f.getChildFilterCount() == 0 &&
+            f.getLhs().getValueCase() == Expression.ValueCase.COLUMNIDENTIFIER)
+        .filter(
+            f -> viewFilterMap.containsKey(f.getLhs().getColumnIdentifier().getColumnName()))
+        .collect(Collectors
+            .toMap(f -> f.getLhs().getColumnIdentifier().getColumnName(), f -> f));
+    if (childLeafFilterMap.isEmpty()) {
+      long matchCount = filter.getChildFilterList().stream()
```

Comment:
If we flip the operator check, we can make this clearer by not using a match count but instead explicit predicates:

```
if (filter.getOperator() == Operator.AND) {
  return filter.getChildFilterList().stream().anyMatch(f -> doViewFiltersMatch(f, viewFilterMap));
} else if (filterOperator() === Operator.OR) {
  return filter.getChildFilterList().stream().allMatch(f -> doViewFiltersMatch(f, viewFilterMap));
}
```
(alternatively, drop the map above and use that in